graphql_compiler/compiler/compiler_frontend.py

```diff
@@ -579,6 +578,9 @@ def _compile_fragment_ast(schema, current_schema_type, ast, location, context):
         current_schema_type.is_same_type(equivalent_union_type)
     )
 
+    if not is_same_type_as_scope:
+        query_metadata_table.record_coercion_at_location(location, coerces_to_type_obj)
```

Comment:
This is fine as a temporary fix, but I'd prefer in the long run to only record coercions if there is actually a `CoerceType` block there. I think this sort of side-steps the problem rather than actually solving it -- there may be another (more complex) edge case that can trigger the same issue.